### PR TITLE
Take absolute values when doing DNS jitter.

### PR DESCRIPTION
**Pull request by Stevenjin8**
Commit Message: Take absolute values when doing DNS jitter.
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #36943
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

In response to #36943. Jitter is generated by taking a random value from the random generator and creating a std::chrono::millisecond.
Since the random value is a uint64, I figured that the resulting std::chrono::millisecond would have to be positive.
However, std::chrono::millisecond uses a int64 under the hood, so large random values were being converted to negative numbers.
For example, if the random generator returned 0x8000000000000000, there would be a jitter of -9223372036854775808ms.
This jitter would in turn case the refresh timer to be negative.

### CODE REVIEW

File: source/extensions/clusters/logical_dns/logical_dns_cluster.cc [hunk 1]
```diff
@@ -152,7 +152,13 @@ void LogicalDnsCluster::startResolve() {
...
           }
           if (dns_jitter_ms_.count() != 0) {
-            final_refresh_rate += std::chrono::milliseconds(random_.random()) % dns_jitter_ms_;
+            // We need to wrap with std::chrono::abs because
+            // random() returns an uint64 that is interpreted as a long by
+            // std::chrono::milliseconds. If the random() return value has a leading 1 in its binary
+            // representation, then it will be interpreted as a negative long, potentially causing
+            // final_refresh_rate to be negative. This causes Envoy to crash.
+            final_refresh_rate +=
+                std::chrono::abs(std::chrono::milliseconds(random_.random()) % dns_jitter_ms_);
```
Comment (KBaichoo) [inline source/extensions/clusters/logical_dns/logical_dns_cluster.cc:161]: I think another way to solve this is to use:
`std::chrono::milliseconds(random_.random() % dns_jitter_ms_.count())` dns_jitter_ms is in the range of valid ms so I think the uint % long long will be a smaller positive number in the range of ms.

This _might_ still be possible to overflow in adding the two if the values are very large e.g. `std::numeric_limit<int64>::max()`

**Review by KBaichoo — COMMENTED** [review]
mostly looks good, thank you

File: source/extensions/clusters/logical_dns/logical_dns_cluster.cc (hunk 1 above)
Comment (Stevenjin8) [reply source/extensions/clusters/logical_dns/logical_dns_cluster.cc:161]: I think `count()` returns an int64, but I agree.

Comment (KBaichoo) [reply source/extensions/clusters/logical_dns/logical_dns_cluster.cc:161]: To clarify the latter end of my comment is it still possible for final_refresh_rate to overflow from the addition?

Comment (Stevenjin8) [reply source/extensions/clusters/logical_dns/logical_dns_cluster.cc:161]: yes, BUT only if `dns_jitter_ms` is negative. I will add a validator and test case for this.

Comment (Stevenjin8) [reply source/extensions/clusters/logical_dns/logical_dns_cluster.cc]: @KBaichoo 
So I'm fairly confident that `std::chrono::milliseconds(random_.random() % dns_jitter_ms_.count())` wont have any overflow shenanigans. The biggest time value we have have in protobufs is 315,576,000,000, so the most milliseconds that can be represented is 31,557,600,000,000). Then `0 < random_.random() % dns_jitter_ms_.count() < 31,557,600,000,000`.
`std::chrono::milliseconds` is represented internally by an int64, which can take values up to 9,223,372,036,854,775,807. So we should be good.  But,

a) would appreciate a check on my logic
b) kinda want to wrap it all in an `abs` and not think about binary representations/limits.

**Comment by Stevenjin8** [comment]
@KBaichoo Bump

File: changelogs/current.yaml [hunk 2]
```diff
@@ -87,6 +87,9 @@ bug_fixes:
   change: |
     Relaxed the restriction on SNI logging to allow the ``_`` character, even if
     ``envoy.reloadable_features.sanitize_sni_in_access_log`` is enabled.
+- area: DNS
+  change: |
+    Fixed bug where setting ``dns_jitter <envoy_v3_api_field_config.cluster.v3.Cluster.dns_jitter>`` to large values caused Envoy to crash.
```
Comment (KBaichoo) [inline changelogs/current.yaml:92]: caused Envoy Bug to fire.


It only aborts in debug mode

File: source/extensions/clusters/logical_dns/logical_dns_cluster.cc (hunk 1 above)
Comment (KBaichoo) [reply source/extensions/clusters/logical_dns/logical_dns_cluster.cc]: Thank you for thinking through this, this lgtm.

**Review by KBaichoo — APPROVED** [review]
Otherwise lgtm, thank you!

**Review by KBaichoo — APPROVED** [review]

**Pull request merged**